### PR TITLE
feat(rdb-pn): add rdb private network support

### PR DESCRIPTION
**Pull request by Monitob**

### CODE REVIEW

File: go.sum [hunk 1]
```diff
@@ -286,6 +286,8 @@ github.com/prometheus/client_model v0.0.0-20190812154241-14fe0d1b01d4/go.mod h1:
 github.com/rogpeppe/go-internal v1.3.0/go.mod h1:M8bDsm7K2OlrFYOpmOWEs/qY81heoFRclV5y23lUDJ4=
 github.com/scaleway/scaleway-sdk-go v1.0.0-beta.7.0.20211005125957-db499d89a262 h1:ioN4zmp/4AoCJ/BbuwmkpH+bnSSntgWMUN1PkQ46/nI=
 github.com/scaleway/scaleway-sdk-go v1.0.0-beta.7.0.20211005125957-db499d89a262/go.mod h1:NnxccdAsg2lS0ng/kfLVLemUQR50FCM/cdZv2gqKDhw=
+github.com/scaleway/scaleway-sdk-go v1.0.0-beta.7.0.20211029161737-75850b6c00c9 h1:eZyNrvcs179sBScC+L6kBHpM+acPbpT5hDMhda7JCrk=
```
Comment (remyleone) [inline go.sum:289]: Could you do a `go mod tidy`?

File: scaleway/data_source_rdb_acl_test.go [hunk 2]
```diff
@@ -10,32 +10,27 @@ import (
...
+	instanceName := "data-source-rdb-acl-basic"
 	resource.ParallelTest(t, resource.TestCase{
 		PreCheck:          func() { testAccPreCheck(t) },
 		ProviderFactories: tt.ProviderFactories,
 		CheckDestroy:      testAccCheckScalewayRdbInstanceDestroy(tt),
 		Steps: []resource.TestStep{
 			{
 				Config: fmt.Sprintf(`
-					resource "scaleway_rdb_instance" "main" {
+					resource "scaleway_rdb_instance" "rdbAcl" {
```
Comment (remyleone) [inline scaleway/data_source_rdb_acl_test.go:21]: why this rename is needed? We already know that it is rdbacl because of the name of the test. Besides, why is acl deleted?

File: scaleway/data_source_rdb_acl_test.go [hunk 3]
```diff
@@ -44,34 +39,41 @@ func TestAccScalewayDataSourceRDBAcl_Basic(t *testing.T) {
...
-						}
 					}
 					data "scaleway_rdb_acl" "maindata" {
-						instance_id = scaleway_rdb_instance.main.id
+						instance_id = scaleway_rdb_instance.rdbAcl.id
 
 					}`, instanceName),
 				Check: resource.ComposeTestCheckFunc(
 					resource.TestCheckResourceAttr("scaleway_rdb_acl.main", "acl_rules.0.ip", "1.2.3.4/32"),
 					resource.TestCheckResourceAttr("data.scaleway_rdb_acl.maindata", "acl_rules.0.ip", "1.2.3.4/32"),
```
Comment (remyleone) [inline scaleway/data_source_rdb_acl_test.go:62]: could you also keep the second acl_rules block and check if the 4.5.6.7 is also present?

File: scaleway/resource_rdb_instance.go [hunk 4]
```diff
@@ -154,7 +181,11 @@ func resourceScalewayRdbInstance() *schema.Resource {
 				Computed:    true,
 				Description: "Certificate of the database instance",
 			},
-
+			"load_balancer": {
+				Type:        schema.TypeBool,
+				Computed:    true,
+				Description: "Load balancers to forward the traffic to the right node based on the node state.",
```
Comment (remyleone) [inline scaleway/resource_rdb_instance.go:187]: I don't understand this sentence. Is there several load balancers?

File: scaleway/resource_rdb_instance_test.go [hunk 5]
```diff
@@ -49,9 +49,9 @@ func TestAccScalewayRdbInstance_Basic(t *testing.T) {
 		Steps: []resource.TestStep{
 			{
 				Config: `
-					resource scaleway_rdb_instance main {
-						name = "test-rdb"
-						node_type = "db-dev-s"
+					resource scaleway_rdb_instance rdbBasic {
```
Comment (remyleone) [inline scaleway/resource_rdb_instance_test.go:52]: Why is this rename needed?

Comment (remyleone) [reply scaleway/resource_rdb_instance_test.go:52]: If you don't have several resources, just naming them `main` or something generic is enough.

File: scaleway/resource_rdb_instance_test.go [hunk 6]
```diff
@@ -263,34 +283,6 @@ func testAccCheckScalewayRdbExists(tt *TestTools, n string) resource.TestCheckFu
 	}
 }
 
-func TestAccScalewayRdbInstance_Capitalize(t *testing.T) {
```
Comment (remyleone) [inline scaleway/resource_rdb_instance_test.go:266]: Why is this test deleted?

File: scaleway/resource_rdb_user_test.go [hunk 7]
```diff
@@ -19,7 +19,7 @@ func init() {
 func TestAccScalewayRdbUser_Basic(t *testing.T) {
 	tt := NewTestTools(t)
 	defer tt.Cleanup()
-	instanceName := "TestAccScalewayRdbUser_Basic"
```
Comment (remyleone) [inline scaleway/resource_rdb_user_test.go:22]: Why this rename is needed?

File: scaleway/resource_rdb_instance_test.go [hunk 8]
```diff
@@ -194,44 +249,9 @@ func TestAccScalewayRdbInstance_Volume(t *testing.T) {
...
-					testAccCheckScalewayRdbExists(tt, "scaleway_rdb_instance.main"),
-					resource.TestCheckResourceAttr("scaleway_rdb_instance.main", "volume_type", "bssd"),
-					resource.TestCheckResourceAttr("scaleway_rdb_instance.main", "volume_size_in_gb", "10"),
-				),
-			},
-		},
-	})
-}
-
-func TestAccScalewayRdbInstance_BackupSchedule(t *testing.T) {
```
Comment (remyleone) [inline scaleway/resource_rdb_instance_test.go:206]: Why is this test deleted?

File: scaleway/resource_rdb_instance_test.go [hunk 9]
```diff
@@ -118,32 +150,55 @@ func TestAccScalewayRdbInstance_Settings(t *testing.T) {
...
+
+					resource scaleway_rdb_instance rdbPN {
+						name = "test-rdb-with-pn"
 						node_type = "db-dev-s"
 						engine = "PostgreSQL-11"
 						is_ha_cluster = false
 						disable_backup = true
 						user_name = "my_initial_user"
 						password = "thiZ_is_v&ry_s3cret"
-						settings = {
```
Comment (remyleone) [inline scaleway/resource_rdb_instance_test.go:129]: Why are the test of settings setting removed? Could it be migrated to a different test?

File: scaleway/resource_rdb_instance_test.go [hunk 10]
```diff
@@ -160,8 +215,8 @@ func TestAccScalewayRdbInstance_Volume(t *testing.T) {
 		Steps: []resource.TestStep{
 			{
 				Config: `
-					resource scaleway_rdb_instance main {
-						name = "test-rdb"
+					resource scaleway_rdb_instance rdbVolume {
```
Comment (remyleone) [inline scaleway/resource_rdb_instance_test.go:218]: why is this change needed?

File: scaleway/resource_rdb_instance_test.go (hunk 6 above)
Comment (Monitob) [reply scaleway/resource_rdb_instance_test.go:266]: This test was integrated in the basic test

File: scaleway/resource_rdb_user_test.go (hunk 7 above)
Comment (Monitob) [reply scaleway/resource_rdb_user_test.go:22]: I was bizarre a camel case naming in the console

File: scaleway/resource_rdb_instance.go [hunk 11]
```diff
@@ -114,6 +118,29 @@ func resourceScalewayRdbInstance() *schema.Resource {
...
+				Optional:    true,
+				MaxItems:    1,
+				Description: "List of private network to expose your database instance",
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"ip": {
+							Type:         schema.TypeString,
+							Required:     true,
+							ValidateFunc: validation.IsCIDR,
+							Description:  "The IP of your service",
```
Comment (remyleone) [inline scaleway/resource_rdb_instance.go:132]: What if we have a DHCP configuration happening in the private network. Range wise, is there tests on rdb side to verify that the IP range we ask is contained in the one allowed by the private network?

File: scaleway/helpers_rdb_test.go [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+		Updates   []interface{}
+		Expected  map[string]bool
+	}{
+		{
+			name: "removeAll",
+			Endpoints: []*rdb.Endpoint{&rdb.Endpoint{
+				ID: "6ba7b810-9dad-11d1-80b4-00c04fd430c1",
+				PrivateNetwork: &rdb.EndpointPrivateNetworkDetails{PrivateNetworkID: "6ba7b810-9dad-11d1-80b4-00c04fd430c8",
+					ServiceIP: scw.IPNet{IPNet: net.IPNet{
+						IP: net.IPv4(0x1, 0x1, 0x1, 0x1), Mask: net.IPMask{0xff, 0xff, 0xff, 0xff}}},
```
Comment (remyleone) [inline scaleway/helpers_rdb_test.go:72]: In all your tests, could you use integer notation which is easier to read: https://pkg.go.dev/net#example-IPv4

File: scaleway/helpers_rdb_test.go [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+					Zone: scw.ZoneFrPar1},
+			}},
+			Expected: map[string]bool{
+				"6ba7b810-9dad-11d1-80b4-00c04fd430c1": true,
+			},
+		},
+		{
+			name: "shouldUpdatePrivateNetwork",
```
Comment (remyleone) [inline scaleway/helpers_rdb_test.go:80]: Could you add a test where we check if we should add a private network at a place where none existed before?

File: scaleway/helpers_rdb_test.go [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+	tests := []struct {
+		name     string
+		A        *rdb.EndpointPrivateNetworkDetails
+		B        *rdb.EndpointPrivateNetworkDetails
+		expected bool
+	}{
+		{
+			name: "isEqualPrivateNetworkDetails",
+			A: &rdb.EndpointPrivateNetworkDetails{PrivateNetworkID: "6ba7b810-9dad-11d1-80b4-00c04fd430c8", ServiceIP: scw.IPNet{IPNet: net.IPNet{
+				IP: net.IPv4(0x1, 0x1, 0x1, 0x1), Mask: net.IPMask{0xff, 0xff, 0xff, 0xff}}}, Zone: scw.ZoneFrPar1},
```
Comment (remyleone) [inline scaleway/helpers_rdb_test.go:22]: Use integer notation and have a field per line, it would be easier to read and would allow to have an helpful diff

File: scaleway/helpers_rdb.go [hunk 15]
```diff
@@ -71,3 +72,101 @@ func expandInstanceSettings(i interface{}) []*rdb.InstanceSetting {
...
+	var res []*rdb.EndpointSpec
+
+	res = append(res, &rdb.EndpointSpec{
+		LoadBalancer: &rdb.EndpointSpecLoadBalancer{}})
+
+	return res
+}
+
+func getRdbEndpointsToRemove(endPoints []*rdb.Endpoint, updates interface{}) (map[string]bool, error) {
+	actions := make(map[string]bool)
```
Comment (remyleone) [inline scaleway/helpers_rdb.go:106]: Could you rename actions in endpointsToRemove ?

File: scaleway/helpers_rdb.go [hunk 16]
```diff
@@ -71,3 +72,101 @@ func expandInstanceSettings(i interface{}) []*rdb.InstanceSetting {
...
+			ServiceIP:        expandIPNet(r["ip"].(string)),
+			Zone:             scw.Zone(locality),
+		}
+
+		pnZonedID := newZonedIDString(pn.Zone, pn.PrivateNetworkID)
+		if _, configExist := configs[pnZonedID]; configExist {
+			isEqual := isEndPointEqual(configs[pnZonedID], pn)
+			// match the endpoint id for a private network
+			if _, endPointExist := privateIDToEndpointID[pnZonedID]; endPointExist {
+				actions[privateIDToEndpointID[pnZonedID]] = !isEqual
```
Comment (remyleone) [inline scaleway/helpers_rdb.go:137]: A bit difficult to read and understand. Could you make it simpler to follow? We are at 3 imbrications levels. Could you give a more explicit name to configs for instance, is it a local config or a remote configuration?

File: scaleway/resource_rdb_instance.go [hunk 17]
```diff
@@ -303,6 +342,15 @@ func resourceScalewayRdbInstanceRead(ctx context.Context, d *schema.ResourceData
 	// set settings
 	_ = d.Set("settings", flattenInstanceSettings(res.Settings))
 
+	// set endpoints
+	hasLoadBalancerEndpoint := true
```
Comment (remyleone) [inline scaleway/resource_rdb_instance.go:346]: Could an instance have neither load balancer nor private networks?

File: scaleway/resource_rdb_instance_test.go [hunk 18]
```diff
@@ -72,43 +72,75 @@ func TestAccScalewayRdbInstance_Basic(t *testing.T) {
...
 				`,
 				Check: resource.ComposeTestCheckFunc(
 					testAccCheckScalewayRdbExists(tt, "scaleway_rdb_instance.main"),
-					resource.TestCheckResourceAttr("scaleway_rdb_instance.main", "name", "test-rdb"),
+					resource.TestCheckResourceAttr("scaleway_rdb_instance.main", "name", "test-rdb-with-cluster"),
 					resource.TestCheckResourceAttr("scaleway_rdb_instance.main", "node_type", "db-dev-m"),
 					resource.TestCheckResourceAttr("scaleway_rdb_instance.main", "engine", "PostgreSQL-11"),
 					resource.TestCheckResourceAttr("scaleway_rdb_instance.main", "is_ha_cluster", "true"),
 					resource.TestCheckResourceAttr("scaleway_rdb_instance.main", "disable_backup", "false"),
+					resource.TestCheckResourceAttr("scaleway_rdb_instance.main", "backup_schedule_frequency", "24"),
```
Comment (remyleone) [inline scaleway/resource_rdb_instance_test.go:124]: Where are those settings defined?

Comment (remyleone) [reply scaleway/resource_rdb_instance_test.go:124]: If they are not defined in the terraform block they shouldn't be tested here.

File: scaleway/resource_rdb_instance_test.go [hunk 19]
```diff
@@ -187,23 +246,39 @@ func TestAccScalewayRdbInstance_Volume(t *testing.T) {
...
+			{
+				Config: `
+					resource scaleway_vpc_private_network pn01 {
+						name = "my_private_network_without_attachment"
+						tags = ["tag0", "tag1", "rdb_pn"]
+					}
+
+					resource scaleway_vpc_private_network pn02 {
+						name = "my_private_network_without_attachment"
+						tags = ["tag2", "tag3"]
```
Comment (remyleone) [inline scaleway/resource_rdb_instance_test.go:269]: If you don't test tags just remove them.

File: scaleway/resource_rdb_instance.go (hunk 11 above)
Comment (remyleone) [reply scaleway/resource_rdb_instance.go:132]: A test that tests how I can combine DHCP and IP settings should be added to test that it works. Adding it to the documentation should be useful too as it presents an interesting use case.

**Review by remyleone — APPROVED** [review]
